### PR TITLE
Throwing error in App if needed product data is missing

### PR DESCRIPTION
**Pull request by ashleemboyer**
## "Ready for Review" checklist

- [x] The Vercel preview link has been added to this PR's description
- [x] The Asana task has been added to this PR's description
- [x] This PR's link has been added to the "📐 Pull Request" field on the Asana task
- [x] This Vercel preview link has been added to the "🔍 Preview" field on the Asana task
- [x] The description template has been filled in below

---

## Relevant links

- [Preview link](https://dev-portal-git-ambadd-check-for-product-hydration-hashicorp.vercel.app/) 🔎
- [Asana task](https://app.asana.com/0/1201979821989355/1201980530846413/f) 🎟️

## What

- Adds a check in `_app.tsx` on DevDot pages that need product data hydrated.

## Why

- I was going to write this out in a README or something, but decided an error might be more effective on the fly. This will help remind us to include a `product` property in every DevDot page component's props. Currently, it's required for `ProductSwitcher`, but some changes will come soon to the main header navigation and more properties in that object will be required.

## How

- Mainly needed to figure out which pages to throw this error on, so I started by just checking if the current page was `/`, then worked through main page paths one by one. That led to getting rid of `/_proxied-dot-io` and `/swingset` pages from the check.

## Testing

There's nothing to manually test here. If there were pages that were failing the check, we would see the build checks below failing.

## Anything else?

Nope!

### CODE REVIEW

File: src/pages/_app.tsx [hunk 1]
```diff
@@ -33,6 +34,23 @@ const { ConsentManager, openConsentManager } = createConsentManager({
...
+  const isDevDotPage =
+    !currentPath.startsWith(`/_proxied-dot-io`) &&
+    !currentPath.startsWith(`/swingset`)
+  const isNotHomePage = currentPath !== '/'
+  const isMissingProductData = !pageProps.product
+  if (isDevDotPage && isNotHomePage && isMissingProductData) {
+    throw new Error(
+      `\`product\` property is missing from \`pageProps\` prop on path: \`${currentPath}\`. Product data is needed to hydrate \`CurrentProductContext\`.`
+    )
+  }
```
Comment (BRKalow) [inline src/pages/_app.tsx:52]: what do you think about gating this behind `if (process.env.HASHI_ENV !== 'production')`? Realistically we shouldn't need to do this check in production once the app has been built and the pages have been rendered.

File: src/pages/_app.tsx [hunk 2]
```diff
@@ -33,6 +34,23 @@ const { ConsentManager, openConsentManager } = createConsentManager({
...
+  const currentPath = useCurrentPath({ excludeHash: true, excludeSearch: true })
+  const isDevDotPage =
+    !currentPath.startsWith(`/_proxied-dot-io`) &&
+    !currentPath.startsWith(`/swingset`)
+  const isNotHomePage = currentPath !== '/'
+  const isMissingProductData = !pageProps.product
+  if (isDevDotPage && isNotHomePage && isMissingProductData) {
+    throw new Error(
+      `\`product\` property is missing from \`pageProps\` prop on path: \`${currentPath}\`. Product data is needed to hydrate \`CurrentProductContext\`.`
+    )
```
Comment (BRKalow) [inline src/pages/_app.tsx:51]: Suggestion: Where can we get the data if it's missing? Maybe mention that the data can be read from `src/data/{product}.json`

**Review by BRKalow — CHANGES_REQUESTED** [review]
Good idea!

File: src/pages/_app.tsx (hunk 1 above)
Comment (ashleemboyer) [reply src/pages/_app.tsx:52]: @BRKalow do you know what `HASHI_ENV` is during the Vercel deploy check? I think it's `'preview'`, but I'm not 100% sure. If it is `'preview'`, I think this'll be a good add!

File: src/pages/_app.tsx (hunk 2 above)
Comment (ashleemboyer) [reply src/pages/_app.tsx:51]: Great idea! I'll put that in the message.

File: src/pages/_app.tsx (hunk 1 above)
Comment (ashleemboyer) [reply src/pages/_app.tsx]: Testing it out with: https://github.com/hashicorp/dev-portal/pull/222/commits/b1ea71ac69ec0f33c4092cb4dc2d751ac72348c6. Will fix the line afterwards. 😊

File: src/pages/_app.tsx (hunk 2 above)
Comment (ashleemboyer) [reply src/pages/_app.tsx]: Updated in: https://github.com/hashicorp/dev-portal/pull/222/commits/346090bf5a9349344b7bb6428d79231d31cd51ba

File: src/pages/_app.tsx (hunk 1 above)
Comment (BRKalow) [reply src/pages/_app.tsx]: Yep, it is `preview` for deploy previews, production builds on main have `production` 👍

**Review by BRKalow — APPROVED** [review]

File: src/pages/_app.tsx [hunk 3]
```diff
@@ -27,6 +28,25 @@ if (typeof window !== 'undefined' && process.env.AXE_ENABLED) {
...
+
+  /**
+   * TODO: this will be temporary and last for as long as we have to manually
+   * hydrate product data from page components. Ideally, we won't have to do
+   * this we use more dynamic routes.
+   */
+  if (process.env.HASHI_ENV !== 'production') {
+    const isDevDotPage =
+      !currentPath.startsWith(`/_proxied-dot-io`) &&
+      !currentPath.startsWith(`/swingset`)
```
Comment (BRKalow) [inline src/pages/_app.tsx:41]: @ashleemboyer I don't think `currentPath` will include `/_proxied-dot-io` when we apply the rewrites, so this check fires when viewing the preview in the context of an io site.

**Pull request merged**